File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/execution/JobProcessor.scala
package org.enso.interpreter.instrument.execution

import org.enso.interpreter.instrument.job.{BackgroundJob, Job}

import scala.concurrent.Future

/** Defines a uniform interface to execute job.
  */
trait JobProcessor {

  /** Runs a job with the provided context.
    *
    * @param job a job to execute
    * @return the future result of an asynchronous computation
    */
  def run[A](job: Job[A]): Future[A]

  /** Runs a job with the provided context in the background.
    *
    * @param job a job to execute
    * @return the future result of an asynchronous computation
    */
  def runBackground[A](job: BackgroundJob[A]): Unit

  /** Stops the job processor. */
  def stop(): Unit

}

File: engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/execution/JobProcessor.java
package org.enso.interpreter.instrument.execution;

import org.enso.interpreter.instrument.job.BackgroundJob;
import org.enso.interpreter.instrument.job.Job;

public interface JobProcessor {

  /**
   * Runs a job with the provided context.
   *
   * @param job a job to execute
   * @return the future result of an asynchronous computation
   */
  <T> scala.concurrent.Future<T> run(Job<T> job);

  /**
   * Runs a job with the provided context in the background.
   *
   * @param job a job to execute
   * @return the future result of an asynchronous computation
   */
  <T> void runBackground(BackgroundJob<T> job);

  /** Stops the job processor. */
  void stop();
}
